File: server/agent.ts
import { AIMessage, HumanMessage, SystemMessage, type BaseMessage } from '@langchain/core/messages';
import type { BaseChatModel } from '@langchain/core/language_models/chat_models';
import { createReactAgent } from '@langchain/langgraph/prebuilt';
import { ChatAnthropic } from '@langchain/anthropic';
import { ChatOpenAI } from '@langchain/openai';
import { ChatGoogleGenerativeAI } from '@langchain/google-genai';
import { ChatMistralAI } from '@langchain/mistralai';
import { ChatGroq } from '@langchain/groq';
import { ChatCohere } from '@langchain/cohere';
import { ChatXAI } from '@langchain/xai';
import { ChatDeepSeek } from '@langchain/deepseek';
import { SUMMARIZE_PROMPT, summaryFallback } from '@deku/core';
import type { ApiKeys, ChatEvent, Msg } from '@deku/core';
import { PROVIDERS, SYSTEM_PROMPT } from './config.js';
import { toolEventFor, toolsFor } from './tools.js';

export function buildModel(provider: string, model: string, apiKeys: ApiKeys = {}): BaseChatModel {
  const def = PROVIDERS.find((p) => p.id === provider);
  if (!def) throw new Error(`Unknown provider: ${provider}`);
  // browser-held key wins; server .env is the fallback
  const apiKey = apiKeys[def.envKey] || process.env[def.envKey];
  if (!apiKey) {
    throw new Error(`No API key for ${def.name} — right-click the avatar → “API keys…” to add one.`);
  }
  switch (provider) {
    case 'anthropic':
      return new ChatAnthropic({ model, apiKey, maxTokens: 4096 });
    case 'openai':
      return new ChatOpenAI({ model, apiKey });
    case 'google':
      return new ChatGoogleGenerativeAI({ model, apiKey });
    case 'mistral':
      return new ChatMistralAI({ model, apiKey });
    case 'groq':
      return new ChatGroq({ model, apiKey });
    case 'cohere':
      return new ChatCohere({ model, apiKey });
    case 'xai':
      return new ChatXAI({ model, apiKey });
    case 'deepseek':
      return new ChatDeepSeek({ model, apiKey });
    default:
      throw new Error(`Unknown provider: ${provider}`);
  }
}

export function toLcMessages(history: Msg[]): BaseMessage[] {
  const out: BaseMessage[] = [new SystemMessage(SYSTEM_PROMPT)];
  for (const m of history) {
    if (m.kind === 'summary') out.push(new AIMessage(`[Summary of earlier conversation] ${m.content}`));
    else if (m.role === 'user') out.push(new HumanMessage(m.content));
    else out.push(new AIMessage(m.content));
  }
  return out;
}

function chunkText(content: unknown): string {
  if (typeof content === 'string') return content;
  if (Array.isArray(content)) {
    return content
      .map((c) => (typeof c === 'string' ? c : c?.type === 'text' || c?.type === 'text_delta' ? (c.text ?? '') : ''))
      .join('');
  }
  return '';
}

/** Stream the agent run as ChatEvents (delta / tool). */
export async function* runAgent(opts: {
  provider: string;
  model: string;
  capabilities: string[];
  history: Msg[];
  apiKeys?: ApiKeys;
  signal?: AbortSignal;
}): AsyncGenerator<Extract<ChatEvent, { type: 'delta' | 'tool' }>> {
  const llm = buildModel(opts.provider, opts.model, opts.apiKeys);
  const agent = createReactAgent({ llm, tools: toolsFor(opts.capabilities, opts.apiKeys) });

  const stream = agent.streamEvents(
    { messages: toLcMessages(opts.history) },
    { version: 'v2', signal: opts.signal },
  );
  for await (const ev of stream) {
    if (ev.event === 'on_chat_model_stream') {
      const text = chunkText(ev.data?.chunk?.content);
      if (text) yield { type: 'delta', text };
    } else if (ev.event === 'on_tool_end') {
      const output = ev.data?.output;
      const outputText = typeof output === 'string' ? output : chunkText(output?.content) || undefined;
      yield { type: 'tool', event: toolEventFor(ev.name, ev.data?.input, outputText) };
    }
  }
}

export async function summarizeMessages(provider: string, model: string, msgs: Msg[], apiKeys: ApiKeys = {}): Promise<string> {
  const transcript = msgs
    .map((m) => `${m.role === 'user' ? 'User' : 'Assistant'}: ${m.content}`)
    .join('\n\n');
  try {
    const llm = buildModel(provider, model, apiKeys);
    const res = await llm.invoke([new SystemMessage(SUMMARIZE_PROMPT), new HumanMessage(transcript)]);
    return chunkText(res.content).trim() || summaryFallback(msgs);
  } catch {
    return summaryFallback(msgs);
  }
}

File: apps/fullstack/backend/deku_server/agent.py
"""Port of server/agent.ts onto langgraph-python + init_chat_model."""
import os
from collections.abc import AsyncIterator

from langchain.chat_models import init_chat_model
from langchain_core.messages import AIMessage, HumanMessage, SystemMessage
from langgraph.prebuilt import create_react_agent

from . import ops
from .config import SUMMARIZE_PROMPT, SYSTEM_PROMPT, provider_by_id
from .models import Msg, dump
from .tools import tool_event_for, tools_for


def build_model(provider: str, model: str, api_keys: dict[str, str]):
    d = provider_by_id(provider)
    if d is None:
        raise ValueError(f"Unknown provider: {provider}")
    # browser-held key wins; server env is the fallback
    key = api_keys.get(d["envKey"]) or os.environ.get(d["envKey"])
    if not key:
        raise ValueError(f"No API key for {d['name']} — right-click the avatar → “API keys…” to add one.")
    kwargs: dict = {"api_key": key}
    if provider == "anthropic":
        kwargs["max_tokens"] = 4096
    return init_chat_model(model, model_provider=d["prefix"], **kwargs)


def to_lc_messages(history: list[Msg]):
    out = [SystemMessage(SYSTEM_PROMPT)]
    for m in history:
        if m.kind == "summary":
            out.append(AIMessage(f"[Summary of earlier conversation] {m.content}"))
        elif m.role == "user":
            out.append(HumanMessage(m.content))
        else:
            out.append(AIMessage(m.content))
    return out


def chunk_text(content) -> str:
    if isinstance(content, str):
        return content
    if isinstance(content, list):
        parts = []
        for c in content:
            if isinstance(c, str):
                parts.append(c)
            elif isinstance(c, dict) and c.get("type") in ("text", "text_delta"):
                parts.append(c.get("text", ""))
        return "".join(parts)
    return ""


async def run_agent(
    provider: str,
    model: str,
    capabilities: list[str],
    history: list[Msg],
    api_keys: dict[str, str],
) -> AsyncIterator[dict]:
    """Yields {'type':'delta','text':...} and {'type':'tool','event':{...}} dicts."""
    llm = build_model(provider, model, api_keys)
    agent = create_react_agent(llm, tools_for(capabilities, api_keys))
    async for ev in agent.astream_events({"messages": to_lc_messages(history)}, version="v2"):
        if ev["event"] == "on_chat_model_stream":
            text = chunk_text(getattr(ev["data"].get("chunk"), "content", None))
            if text:
                yield {"type": "delta", "text": text}
        elif ev["event"] == "on_tool_end":
            output = ev["data"].get("output")
            if isinstance(output, str):
                output_text = output
            else:
                output_text = chunk_text(getattr(output, "content", None)) or None
            event = tool_event_for(ev["name"], ev["data"].get("input"), output_text)
            yield {"type": "tool", "event": dump(event)}


async def summarize_messages(
    provider: str, model: str, msgs: list[Msg], api_keys: dict[str, str]
) -> str:
    transcript = "\n\n".join(
        f"{'User' if m.role == 'user' else 'Assistant'}: {m.content}" for m in msgs
    )
    try:
        llm = build_model(provider, model, api_keys)
        res = await llm.ainvoke([SystemMessage(SUMMARIZE_PROMPT), HumanMessage(transcript)])
        return chunk_text(res.content).strip() or ops.summary_fallback(msgs)
    except Exception:
        return ops.summary_fallback(msgs)
